botcore/utils/scheduling.py

```diff
@@ -0,0 +1,246 @@
+"""Generic python scheduler."""
+
+import asyncio
+import contextlib
+import inspect
+import typing
+from datetime import datetime
+from functools import partial
+
+from botcore.utils import loggers
+
+
+class Scheduler:
+    """
+    Schedule the execution of coroutines and keep track of them.
+
+    When instantiating a Scheduler, a name must be provided. This name is used to distinguish the
+    instance's log messages from other instances. Using the name of the class or module containing
+    the instance is suggested.
+
+    Coroutines can be scheduled immediately with `schedule` or in the future with `schedule_at`
+    or `schedule_later`. A unique ID is required to be given in order to keep track of the
+    resulting Tasks. Any scheduled task can be cancelled prematurely using `cancel` by providing
+    the same ID used to schedule it.  The `in` operator is supported for checking if a task with a
+    given ID is currently scheduled.
+
+    Any exception raised in a scheduled task is logged when the task is done.
+    """
+
+    def __init__(self, name: str):
+        """
+        Initialize a new Scheduler instance.
+
+        Args:
+            name: The name of the scheduler. Used in logging, and namespacing.
+        """
+        self.name = name
+
+        self._log = loggers.get_logger(f"{__name__}.{name}")
+        self._scheduled_tasks: typing.Dict[typing.Hashable, asyncio.Task] = {}
+
+    def __contains__(self, task_id: typing.Hashable) -> bool:
+        """
+        Return True if a task with the given `task_id` is currently scheduled.
```

Comment:
`True` should be sphinx-ified.

Reply:
https://github.com/python-discord/bot-core/pull/29/commits/d7841816581737b4a7cece5bba98328a44630d4b